src/cloudflare/internal/pipeline-transform.ts

```diff
@@ -0,0 +1,132 @@
+    }
+
+    partial = chunk.substring(start, end)
+    start = 0
+    end = 0
+  }
+
+  if (partial.length > 0) {
+    yield partial
+  }
+}
+
+type Batch = {
+  id: string 	// unique identifier for the batch
+  shard?: string 	// assigned shard
+  ts: number 	// timestamp of the event
+
+  format: Format
+  data: unknown
+}
+
+type JsonStream = Batch & {
+  format: Format.JSON_STREAM,
+  data: ReadableStream<Uint8Array>
+}
+
+enum Format {
+  JSON_STREAM = 'json_stream', //jsonl
+}
+
+export class PipelineTransformImpl extends entrypoints.WorkerEntrypoint {
+  #batch?: Batch
+  #initalized: boolean = false
+
+  constructor(ctx: unknown, env: unknown) {
+    super(ctx, env)
+  }
+
+  // stub overriden on the sub class
+  // @ts-ignore
+  public async transformJson(data: object[]): Promise<object[]> {
+    throw new Error('should be implemented by parent')
+  }
+
+  // called by the dispatcher which then calls the subclass methods
+  // @ts-ignore
+  private async _transform(batch: Batch): Promise<JsonStream> {
+    if (this.#initalized) {
+      throw new Error('pipeline entrypoint has already been initialized')
+    }
+
+    // test call to make sure the pipeline entrypoint is setup correctly
+    if (batch === null) {
+      // making sure the function was overriden by an implementing subclass
+      if (this.transformJson !== PipelineTransformImpl.prototype.transformJson) {
+        return null as unknown as JsonStream
+      } else {
+        throw new Error('the transformJson method must be overridden by the PipelineTransform subclass')
+      }
+    } else {
+      this.#batch = batch
+      this.#initalized = true
+
+      switch (this.#batch!.format) {
+        case Format.JSON_STREAM:
+          const data = await this.readJsonStream()
+          const transformed = await this.transformJson(data)
+          return this.sendJson(transformed)
+        default:
+          throw new Error('unsupported batch format')
+      }
+    }
+  }
+
+  private async readJsonStream(): Promise<object[]> {
+    if (this.#batch!.format !== Format.JSON_STREAM) {
+      throw new Error(`expected JSON_STREAM not ${this.#batch!.format}`)
+    }
+
+    const batch = this.#batch!.data as ReadableStream<Uint8Array>
+    const decoder = batch.pipeThrough(new TextDecoderStream())
+
+    const data: object[] = []
+    for await (const line of readLines(decoder)) {
+      data.push(JSON.parse(line))
+    }
+
+    return data
+  }
+
+  private async sendJson(data: object[]): Promise<JsonStream> {
+    const readable = new ReadableStream<string>({
+      start(controller) {
+        for (const obj of data) {
+          const json = `${JSON.stringify(obj)}\n`
+          controller.enqueue(json)
+        }
+
+        controller.close()
+      }
+    })
+
+    return {
+      id: this.#batch!.id,
+      shard: this.#batch!.shard,
+      ts: this.#batch!.ts,
+      format: Format.JSON_STREAM,
+      data: readable.pipeThrough(new TextEncoderStream())
+    } as JsonStream
```

Comment:
Why is this `as JsonStream` required?

Reply:
https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes
I think I would have to declare it as a new variable without shard and then set it after